### PR TITLE
Refactor agent

### PR DESCRIPTION
**Pull request by aviramha**

### CODE REVIEW

File: mirrord-agent/src/sniffer.rs [hunk 1]
```diff
@@ -93,12 +93,73 @@ fn is_closed_connection(flags: u16) -> bool {
     0 != (flags & (TcpFlags::FIN | TcpFlags::RST))
 }
 
-struct ConnectionManager {
-    sessions: SessionMap,
-    index_allocator: IndexAllocator<ConnectionID>,
-    ports: HashSet<u16>,
+pub struct TcpManagerCodec {}
```
Comment (meowjesty) [inline mirrord-agent/src/sniffer.rs:96]: Empty struct could be turned into just a marker struct (also add a bunch of derives to it if it's going to be just a marker).

File: mirrord-agent/src/sniffer.rs [hunk 2]
```diff
@@ -93,12 +93,73 @@ fn is_closed_connection(flags: u16) -> bool {
...
+    capture.set_datalink(Linktype::ETHERNET)?;
+    // Set a dummy filter that shouldn't capture anything. This makes the code easier.
+    capture.filter(DUMMY_BPF, true)?;
+    capture = capture.setnonblock()?;
+
+    Ok(capture)
 }
 
+struct TcpPacketData {
+    data: Vec<u8>,
```
Comment (meowjesty) [inline mirrord-agent/src/sniffer.rs:130]: I would rename this to `bytes`, otherwise I foresee some `let _ = data.data.len()` call.

**Review by meowjesty — COMMENTED** [review]
First look.

File: mirrord-agent/src/sniffer.rs [hunk 3]
```diff
@@ -111,207 +172,343 @@ fn format_bpf(ports: &[u16]) -> String {
...
+#[derive(Debug)]
+pub struct SnifferCommand {
+    client_id: ClientID,
+    command: SnifferCommands,
+}
+
+pub struct TCPSnifferAPI {
+    client_id: ClientID,
+    sender: Sender<SnifferCommand>,
+    enabled: bool,
```
[the remark this answers is not in the document]
Comment (aviramha) [reply mirrord-agent/src/sniffer.rs:193]: How would we benefit from that if we can't implement the variant?

File: mirrord-agent/src/sniffer.rs (hunk 1 above)
Comment (aviramha) [reply mirrord-agent/src/sniffer.rs:96]: Can you elaborate? I'm not sure I follow

Comment (meowjesty) [reply mirrord-agent/src/sniffer.rs:96]: Just saying that if this struct has no fields, drop the `{}` (and feel free to add derive Clone, Debug, and so on).

File: mirrord-agent/src/sniffer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (meowjesty) [reply mirrord-agent/src/sniffer.rs:193]: Then why do we need this bool?

[the remark this answers is not in the document]
Comment (aviramha) [reply mirrord-agent/src/sniffer.rs:194]: It's for the drop..

[the remark this answers is not in the document]
Comment (aviramha) [reply mirrord-agent/src/sniffer.rs]: done

File: mirrord-agent/src/util.rs [hunk 4]
```diff
@@ -75,7 +77,8 @@ where
         }
     }
 
-    /// Removes a topic and all of it's clients
+    // /// Removes a topic and all of it's clients
+    #[allow(dead_code)]
```
Comment (eyalb181) [inline mirrord-agent/src/util.rs:81]: We should maybe document whenever we use allow(dead_code)

File: mirrord-agent/tests/blackbox.rs [hunk 5]
```diff
@@ -26,23 +26,26 @@ mod tests {
...
-            .stdout(Stdio::piped())
-            .stderr(Stdio::piped())
             .spawn()
             .expect("mirrord-agent failed to start");
         // Wait for agent to listen
         sleep(Duration::from_millis(2000)).await;
         let stream = TcpStream::connect("127.0.0.1:61337")
             .await
             .expect("connection to agent failed");
+        println!("..");
```
Comment (eyalb181) [inline mirrord-agent/tests/blackbox.rs:36]: Was this intentionally left in? Don't we want proper logging in tests?

**Review by eyalb181 — CHANGES_REQUESTED** [review]

**Review by eyalb181 — APPROVED** [review]

**Review by meowjesty — COMMENTED** [review]
lgtm +1

**Pull request merged**